R/undesirable_function_linter.R

```diff
@@ -22,12 +22,15 @@ undesirable_function_linter <- function(fun = default_undesirable_functions,
       tokens <- "SYMBOL_FUNCTION_CALL"
     }
 
-    xpath <- sprintf(
-      "//*[(%s) and (%s) and not(parent::expr/preceding-sibling::expr[SYMBOL_FUNCTION_CALL[%s]])]",
+    xpath <- paste0("//*[", xp_and(
       paste0("self::", tokens, collapse = " or "),
```

Comment:
Curious what is faster, `//SYMBOL[...] | //SYMBOL_FUNCTION_CALL[...]` or `//*[[(self::SYMBOL or self::SYMBOL_FUNCTION_CALL) and ...]`.

I'm guessing the first XPath might be faster because it doesn't have to enter all nodes, but not sure.
We have this idiom in a couple of places so may be worth investigating.

Reply:
if there is a "smart" compilation stage, the execution could be the same for both. agree it's worth investigating